src/main/java/com/hierynomus/smbj/share/NamedPipe.java

```diff
@@ -141,7 +196,15 @@ public int transact(byte[] inBuffer, byte[] outBuffer) {
      * @return the number of bytes written to <code>outBuffer</code>
      */
     public int transact(byte[] inBuffer, int inOffset, int inLength, byte[] outBuffer, int outOffset, int outLength) {
-        return ioctl(FSCTL_PIPE_TRANSCEIVE, true, inBuffer, inOffset, inLength, outBuffer, outOffset, outLength);
+        if (outBuffer.length - outOffset < outLength) {
+            throw new IllegalArgumentException();
+        }
+        final byte[] transactOutBuffer = transact(inBuffer);
+        if (transactOutBuffer.length > outLength) {
+            throw new BufferOverflowException();
```

Comment:
Tested against Windows 2008.

Test scenarios:
* namedPipe.write() -> namedPipe.transact() = STATUS_PIPE_BUSY
* namedPipe.write() -> namedPipe.read(zero or partial output buffer) -> namedPipe.transact() = remaining output data from namedPipe.write() plus new data from namedPipe.transact()
* namedPipe.transact(zero or partial output buffer) -> namedPipe.write() -> namedPipe.read() = remaining output data from namedPipe.transact() plus new data from namedPipe.write()

Summary:
If you write() to a NamedPipe you must read() from it at least zero bytes otherwise STATUS_PIPE_BUSY is returned.  If you transact() to a NamedPipe you must read the complete transaction otherwise STATUS_PIPE_BUSY is returned.

The server solves the synchronization problem for us in this case with the exception that you can trick it if you write() and don't read() the entire response.  However, it looks to be perfectly valid to write() multiple operations before calling a single read(); so the synchronization is around transact() but not read() and write().

Based on these tests, the proposals would not protect against synchronization issues.  To fully protect the caller from causing issues with the transact() call, both read() and write() would have to be protected/private methods; but the caller would no longer be able to fully use the NamedPipe API in that case (calling multiple writes before reads).